### PR TITLE
[pkg/translator/azure] New component for translating from Azure native formats

### PR DESCRIPTION
**Pull request by cparkins**
**Description:** <Describe what has changed.>

Adding a Translator to convert from Azure Resource Log format (for now) to OTel.

**Link to tracking Issue:** <Issue number if applicable>
#18210

**Testing:** <Describe what testing was performed and which tests were added.>
Performed manual validation testing between Event Hubs and OTel and added unit tests to cover new use case to allow multiple resources to be included in a batch. Existing tests moved to new package.

**Documentation:** <Describe the documentation added.>
None was necessary as this format is already supported by the Event Hub Receiver.

### CODE REVIEW

**Comment by cparkins** [comment]
@Aneurysm9 - I'd love to get a review of this now that the build is functioning. Next time I'll RTFM much earlier!

**Review by MovieStoreGuy — APPROVED** [review]
This all seems reasonable to me, I will leave it to the code sponsers to have the final approval.

File: pkg/translator/azure/go.mod [hunk 1]
```diff
@@ -0,0 +1,52 @@
+//module github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/azure
+
+go 1.19
```
[the remark this answers is not in the document]
Comment (cparkins) [reply pkg/translator/azure/go.mod:3]: Yes, I'll get that corrected.

**Comment by jpkrohling** [comment]
> I will leave it to the code sponsers to have the final approval.

Is that @atoulme? I think I'm a sponsor for an Azure component, but not sure this here is part of that.

**Comment by cparkins** [comment]
@MovieStoreGuy @jpkrohling I could use some assistance in understanding the build issues I'm running into.  They do not appear to be related to the changes that I made. Is there something that I need to do to resolve these? I've rebase about a dozen times and still get the same error. I just rebased and still get the following from 'make otelcontribcol':

`go: downloading github.com/antonmedv/expr v1.15.2
cd ./cmd/otelcontribcol && GO111MODULE=on CGO_ENABLED=0 go build -trimpath -o ../../bin/otelcontribcol_darwin_arm64 \
		-tags "" .
# github.com/prometheus/prometheus/model/labels
../../../../go/pkg/mod/github.com/prometheus/prometheus@v0.47.0/model/labels/labels.go:365:23: type func(a Label, b Label) bool of func(a, b Label) bool {…} does not match inferred type func(a Label, b Label) int for func(a E, b E) int
../../../../go/pkg/mod/github.com/prometheus/prometheus@v0.47.0/model/labels/labels.go:389:23: type func(a Label, b Label) bool of func(a, b Label) bool {…} does not match inferred type func(a Label, b Label) int for func(a E, b E) int
../../../../go/pkg/mod/github.com/prometheus/prometheus@v0.47.0/model/labels/labels.go:594:24: type func(a Label, b Label) bool of func(a, b Label) bool {…} does not match inferred type func(a Label, b Label) int for func(a E, b E) int
../../../../go/pkg/mod/github.com/prometheus/prometheus@v0.47.0/model/labels/labels.go:621:25: type func(a Label, b Label) bool of func(a, b Label) bool {…} does not match inferred type func(a Label, b Label) int for func(a E, b E) int
# github.com/SAP/go-hdb/driver/internal/protocol/auth
../../../../go/pkg/mod/github.com/!s!a!p/go-hdb@v1.3.10/driver/internal/protocol/auth/auth.go:53:27: type func(a Method, b Method) bool of func(a, b Method) bool {…} does not match inferred type func(a Method, b Method) int for func(a E, b E) int
make: *** [otelcontribcol] Error 1`

You might note that I'm building on MacOS (Darwin) could the issue be related to that??? I'm beyond my ability to solve this problem and don't really know why my changes are causing this issue. Any help would be appreciated.

**Comment by mx-psi** [comment]
@codeboten could you help with the above? I believe you also saw this issue

**Comment by codeboten** [comment]
@mx-psi @cparkins yes it's documented in this issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/24881

There's a link to the underlying prometheus issue that has a PR for it.

File: pkg/translator/azure/go.mod [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+require (
+	github.com/json-iterator/go v1.1.12
+	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatatest v0.85.0
+	github.com/relvacode/iso8601 v1.3.0
+	github.com/stretchr/testify v1.8.4
+	go.opentelemetry.io/collector/component v0.85.0
+	go.opentelemetry.io/collector/pdata v1.0.0-rcv0014
+	go.opentelemetry.io/collector/semconv v0.85.0
+	go.uber.org/zap v1.25.0
+	golang.org/x/exp v0.0.0-20230905200255-921286631fa9
```
Comment (dmitryax) [inline pkg/translator/azure/go.mod:14]: The failing test can be resolved by downgrading this dep to v0.0.0-20230713183714-613f0c0eb8a1 for now. The new version isn't required for this PR

**Comment by cparkins** [comment]
> > I will leave it to the code sponsers to have the final approval.
> 
> Is that @atoulme? I think I'm a sponsor for an Azure component, but not sure this here is part of that.

Yes, @atoulme was set as the code owner because this is code he originally wrote for the Event Hub Receiver.  I would really like to see this get merged as rebasing/updating is becoming more arduous than I have any patience for any longer.  If a someone else would like to become code owner so this can get merged I would appreciate it.  I'm still trying to work through the build issues and assume more changes will be coming once the v0.88 updates are completed. I will make them ASAP and resolve any merge conflicts as I become aware of them.

**Review by atoulme — APPROVED** [review]
Approved!

**Comment by mx-psi** [comment]
@cparkins I can merge this as soon as conflicts are resolved :)

File: .github/CODEOWNERS [hunk 3]
```diff
@@ -144,7 +145,6 @@ pkg/translator/prometheusremotewrite/                                   @open-te
 pkg/translator/signalfx/                                                @open-telemetry/collector-contrib-approvers @dmitryax
 pkg/translator/zipkin/                                                  @open-telemetry/collector-contrib-approvers @MovieStoreGuy @astencel-sumo @crobert-1
 pkg/winperfcounters/                                                    @open-telemetry/collector-contrib-approvers @dashpole @Mrod1598 @BinaryFissionGames
-
```
Comment (atoulme) [inline .github/CODEOWNERS:147]: please undelete this line

File: cmd/otelcontribcol/builder-config.yaml [hunk 4]
```diff
@@ -416,3 +416,6 @@ replaces:
   - github.com/open-telemetry/opentelemetry-collector-contrib/connector/spanmetricsconnector => ../../connector/spanmetricsconnector
   - github.com/openshift/api v3.9.0+incompatible => github.com/openshift/api v0.0.0-20180801171038-322a19404e37
   - github.com/outcaste-io/ristretto v0.2.0 => github.com/outcaste-io/ristretto v0.2.1
+  - github.com/DataDog/datadog-agent/pkg/proto => github.com/DataDog/datadog-agent/pkg/proto v0.48.0-beta.1
+  - github.com/DataDog/datadog-agent/pkg/trace => github.com/DataDog/datadog-agent/pkg/trace v0.48.0-beta.1
```
Comment (atoulme) [inline cmd/otelcontribcol/builder-config.yaml:420]: these replace statements are odd to me, why are they present?

File: pkg/translator/azure/metadata.yaml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+status:
+  codeowners:
+    active: [open-telemetry/collector-approvers, atoulme]
```
Comment (atoulme) [inline pkg/translator/azure/metadata.yaml:3]: If you want to add yourself there, feel free

File: receiver/azureeventhubreceiver/factory.go [hunk 6]
```diff
@@ -21,7 +21,7 @@ const (
 )
 
 var (
-	errUnexpectedConfigurationType = errors.New("Failed to cast configuration to Azure Event Hub Config")
+	errUnexpectedConfigurationType = errors.New("failed to cast configuration to azure event hub config")
```
Comment (atoulme) [inline receiver/azureeventhubreceiver/factory.go:24]: We can do this in a separate PR, ok to leave it in

**Comment by atoulme** [comment]
Sorry, there's just a couple things that caught my eye on second read. Please see my comments. We're very close.

File: cmd/otelcontribcol/builder-config.yaml (hunk 4 above)
Comment (cparkins) [reply cmd/otelcontribcol/builder-config.yaml:420]: I'm not entirely sure. Somewhere along the way the got included, they may have been included in another package.  I can remove them. I've been more focused on fixing the build. I can look at removing that though.

File: pkg/translator/azure/metadata.yaml (hunk 5 above)
Comment (cparkins) [reply pkg/translator/azure/metadata.yaml:3]: Okay, I can do that.

File: receiver/azureeventhubreceiver/factory.go (hunk 6 above)
Comment (cparkins) [reply receiver/azureeventhubreceiver/factory.go:24]: I think my editor may have done this. So I'm just going to leave it for now.

**Comment by cparkins** [comment]
> @cparkins I can merge this as soon as conflicts are resolved :)

@mx-psi Should be ready to go (finally!)

**Comment by cparkins** [comment]
> Sorry, there's just a couple things that caught my eye on second read. Please see my comments. We're very close.

Thanks! I reviewed the final set of changes and made those recommendations.

**Pull request merged**